csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,374 @@
+// This file is part of the ArmoniK project
+//
+// Copyright (C) ANEO, 2022-2023.All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+[PublicAPI]
+public static class ParallelSelectExt
+{
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks.
+  ///   At most `parallelism` tasks will be running at any given time.
+  ///   If `parallelism` is 0, number of threads is used.
+  ///   If `parallelism` is negative, no limit is used.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelism">Maximum number of parallel tasks to be spawned at any moment</param>
+  /// <param name="cancellationToken">Token to cancel the enumeration</param>
+  /// <typeparam name="T">Return type of the tasks</typeparam>
+  /// <returns>Asynchronous results of tasks</returns>
+  [PublicAPI]
+  public static async IAsyncEnumerable<T> ParallelWait<T>(this IEnumerable<Task<T>>                  enumerable,
```

Comment:
Prefer an interface using the C# library elements (https://learn.microsoft.com/fr-fr/dotnet/api/system.threading.tasks.parallel.foreach?view=net-7.0)
Here, TaskOptions could be used.

I also wonder if it would not be nest to have as input a synchronous enumerable (IEnumerable<TInput>) and a second argument providing a Func<TInput, T>.

Reply:
Concerning your first point, that might be a good idea. Will do.

As for your second point, that might just be `ParallelSelect`:
- [`IEnumerable`](https://github.com/aneoconsulting/ArmoniK.Utils/pull/5/files#diff-17b8bc7b72e454fcd9449d12351945a69133bf49d77769c2faf97f3171384633R309-R312)
- [`IAsyncEnumerable`](https://github.com/aneoconsulting/ArmoniK.Utils/pull/5/files#diff-17b8bc7b72e454fcd9449d12351945a69133bf49d77769c2faf97f3171384633R331-R334)

Reply:
@wkirschenmann @aneojgurhem I have added a `ParallelTaskOptions` class. What do you think about this interface?

Reply:
It works pretty well